packages/device-connector/src/arduino-cli/client.ts

```diff
@@ -338,20 +378,77 @@ export class RPCClient {
     })
   }
 
-  getDevices (): Device[] {
-    return this.devices
+  private async addDevice (detectedPort: DetectedPort): Promise<void> {
+    if (detectedPort.port == null) {
+      logger.warn('Port not defined')
+      return
+    }
+
+    if (detectedPort?.matching_boards === undefined || detectedPort.matching_boards.length === 0) {
+      return
+    }
+
+    const board = detectedPort.matching_boards[0]
+    if (board.fqbn == null) {
+      logger.warn('Could not register device: No fqbn found')
+      return
+    }
+
+    const existingDevice = this.devices.find((device) => device.port === detectedPort.port)
+
+    if (existingDevice !== undefined) {
+      if (existingDevice.fqbn === board.fqbn) {
+        return
+      }
+
+      await this.removeDevice(existingDevice)
+    }
+
+    const id = await registerNewDevice(board.fqbn, board.name ?? 'Unknown Device Name')
+
+    const device: Device = {
+      id,
+      name: board.name ?? 'Unknown Device Name',
+      fqbn: board.fqbn,
+      port: detectedPort.port
+    }
+
+    this.devices.push(device)
+    logger.info(`Device attached: ${device.name}`)
+  }
+
+  async initializeDevices (): Promise<void> {
+    const boards = await this.listBoards()
+    for (const detectedPort of boards) {
+      await this.addDevice(detectedPort)
+    }
   }
 
-  setDevices (devices: Device[]): void {
-    this.devices = devices
+  getDevices (): Device[] {
+    return this.devices
   }
 
   async removeDevice (device: Device): Promise<void> {
-    const newDeviceList = this.devices.filter((deviceItem) => deviceItem !== device)
-    this.setDevices(newDeviceList)
+    this.devices = this.devices.filter((deviceItem) => deviceItem !== device)
     setStoredDevices(this.devices)
 
     logger.info(`Device removed: ${device.name}`)
     await deleteDeviceRequest(device.id)
   }
+
+  async removeAllDevices (): Promise<void> {
+    for (const device of this.devices) {
+      try {
+        await deleteDeviceRequest(device.id)
+      } catch {
```

Comment:
Log at least

Reply:
The logging here was removed on purpose. Reason being that this method is called (more so in later commits) in certain cases without needing any output. This function will only ever be called during startup or shutdown. Logging that is not necessary here. I can add a log on shutdown that the function is going to be called however (rather as a debug then though)